src/main/java/org/jabref/gui/externalfiles/ImportHandler.java

```diff
@@ -62,40 +71,96 @@ public ExternalFilesEntryLinker getLinker() {
-                        entriesToAdd = xmpEntriesInFile;
-                    } else {
-                        entriesToAdd = xmpEntriesInFile;
+    public BackgroundTask<List<ImportFilesResultItemViewModel>> importFilesInBackground(List<Path> files) {
+        return new BackgroundTask<>() {
+
+            @Override
+            protected List<ImportFilesResultItemViewModel> call() {
+                results = new ArrayList<>();
+                counter = 1;
+                CompoundEdit ce = new CompoundEdit();
+                for (int i = 0; i < files.size(); i++) {
+
+                    if (isCanceled()) {
+                        break;
                     }
-                } else {
-                    if (!pdfResult.isEmpty()) {
-                        entriesToAdd = pdfResult;
-                    } else {
-                        entriesToAdd = Collections.singletonList(createEmptyEntryWithLink(file));
+                    DefaultTaskExecutor.runInJavaFXThread(() -> {
+                        updateMessage(Localization.lang("Processing file %0 of %1", counter, files.size()));
+                        updateProgress(counter, files.size() - 1);
+                    });
+
+                    var file = files.get(0);
+                    entriesToAdd = Collections.emptyList();
+
+                    try {
+                        if (FileUtil.getFileExtension(file).filter("pdf"::equals).isPresent()) {
+
+                            var pdfImporterResult = contentImporter.importPDFContent(file);
+                            List<BibEntry> pdfEntriesInFile = pdfImporterResult.getDatabase().getEntries();
+
+                            if (pdfImporterResult.hasWarnings()) {
+                                addResultToList(file, false, Localization.lang("Error reading PDF content: %0", pdfImporterResult.getErrorMessage()));
+                            }
+
+                            var xmpParserResult = contentImporter.importXMPContent(file);
+                            List<BibEntry> xmpEntriesInFile = xmpParserResult.getDatabase().getEntries();
+
+                            if (xmpParserResult.hasWarnings()) {
+                                addResultToList(file, false, Localization.lang("Error reading XMP content: %0", xmpParserResult.getErrorMessage()));
+                            }
+
+                            // First try xmp import, if empty try pdf import, otherwise create empty entry
+                            if (!xmpEntriesInFile.isEmpty()) {
+                                entriesToAdd = xmpEntriesInFile;
+                                addResultToList(file, true, Localization.lang("Importing using XMP data..."));
+                            } else if (!pdfEntriesInFile.isEmpty()) {
+                                    entriesToAdd = pdfEntriesInFile;
+                                    addResultToList(file, true, Localization.lang("Importing using extracted PDF data"));
```

Comment:
wouldn't it be good to have a `result` variable, and then add it at the end of the import? Otherwise you might end up with multiple messages for one file nad that can be confusing

Reply:
If you add a new constructor to the ImportFilesResultItemViewModel class, this could be simply `result = new ImportFilesResultItemViewModel(file, xmpParserResult)`.

Reply:
But what if you cancel the task?

Reply:
But I still would need to distinguish each parser result and check if empry or if it  has failed or not. That would be simply duplicating the loop